Add configurable options to the AppEngine object.

rrch opened the pull request:
Use an http.Agent for all api and metadata calls.
Default maxSockets for the agent to 100 (https://github.com/GoogleCloudPlatform/appengine-nodejs/issues/34).

rrch committed:
Add configurable options to the AppEngine object.

Use an http.Agent for all api and metadata calls.
Default maxSockets for the agent to 100.

rakyll commented on lib/index.js:
```diff
@@ -42,11 +42,19 @@ var http = require('http');
   this.cache_ = {};
-  this.logSizeThreshold_ = 1024 * 1024; // 1 MB
-  this.logMaxTimestampDeltaMillis_ = 60000; // 60 seconds
+  this.defaults_ = {
```
Would be lovely if we remove the defaults from the prototype.

rrch replied:
Did you mean "remove from" or "move to"? E.g.
AppEngine.prototype.defaultOptions_ = { foo: 1, bar: 2 };

rakyll commented on lib/index.js:
```diff
@@ -42,11 +42,19 @@ var http = require('http');
+    loggerMaxFileSize: 100 * 1024 * 1024, // 100 MB
+    httpAgentMaxSockets: 100
+  };
+  this.logSizeThreshold_ = this.options_.logSizeThreshold || this.defaults_.logSizeThreshold;
```
We need a function that  overrides the values of the defaults with the values of the options.

Requiring user to rely on `options_.v || defaults_.v` is not great for readability in the longer term.

rrch committed:
Improve code that handles default options.

rrch commented:
PTAL

rrch committed:
Made AppEngine constructor invoke exportAll.

By default. Also added an option to skip that.

This is necessary to make constructing an AppEngine instance with some
options work more naturally.

rrch committed:
Update createAppEngine to avoid calling exportAll twice.

rakyll commented on lib/index.js:
```diff
@@ -58,13 +60,28 @@ function AppEngine() {
+  loggerMaxFiles: 1,
+  loggerMaxFileSize: 100 * 1024 * 1024, // 100 MB
+  httpAgentMaxSockets: 100,
+  exportPublicFunctions: true
```
We should add a comment it's building the surface layer. I'd go with different name such as initialize or build or whatever as we discussed earlier.

rrch replied:
Done. I've renamed it to "addPublicAPIs" and I've also renamed exportAll_ to addPublicAPIs_.

rakyll commented on lib/index.js:
```diff
@@ -986,9 +1006,7 @@ function makeModulesGetHostnameProto(module, version, instance) {
+  return new AppEngine();
 }
 
 module.exports = createAppEngine();
```
module.exports = new AppEngine();

rrch replied:
Done

rakyll commented on lib/index.js:
```diff
@@ -986,9 +1006,7 @@ function makeModulesGetHostnameProto(module, version, instance) {
 }
 
 function createAppEngine() {
```
Remove.

rrch replied:
Done

rakyll commented on test/index.js:
```diff
@@ -49,9 +49,84 @@ describe('appengine', function() {
+        var transport = ae.logger_.transports.file;
+        assert.ok(!!transport);
+        assert.strictEqual(transport.maxFiles, ae.defaultOptions_.loggerMaxFiles);
+        assert.strictEqual(transport.maxsize, ae.defaultOptions_.loggerMaxFileSize);
```
Add an assertion for the exportPublicFunctions.

rrch replied:
Done

rakyll commented on test/index.js:
```diff
@@ -475,6 +550,21 @@ describe('appengine', function() {
 
+      it('uses getRemoteApiRequestOptions_ to get the request options and passes them to sendHttpRequest_', function(done) {
+        var ae = new appengine.AppEngine();
+        var requestOptions = {};
```
Shouldn't we primarily test it with a non-empty object?

rrch replied:
We're using strictEqual in the test, so only the identity of the object matters.

rrch committed:
Rename exportPublicFunctions to addPublicAPIs.

Also renamed exportAll_.

rakyll commented:
LGTM
